llvm/include/llvm/ADT/PagedVector.h

```diff
@@ -0,0 +1,133 @@
+//===- llvm/ADT/PagedVector.h - 'Lazyly allocated' vectors --------*- C++
+//-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the PagedVector class.
+//
+//===----------------------------------------------------------------------===//
+#ifndef LLVM_ADT_PAGEDVECTOR_H
+#define LLVM_ADT_PAGEDVECTOR_H
+
+#include <cassert>
+#include <vector>
+
+namespace llvm {
+// A vector that allocates memory in pages.
+// Order is kept, but memory is allocated only when one element of the page is
+// accessed. This introduces a level of indirection, but it is useful when you
+// have a sparsely initialised vector where the full size is allocated upfront
+// with the default constructor and elements are initialised later, on first
+// access.
+//
+// Notice that this does not have iterators, because if you
+// have iterators it probably means you are going to touch
+// all the memory in any case, so better use a std::vector in
+// the first place.
+template <typename T, int PAGE_SIZE = 1024 / sizeof(T)> class PagedVector {
+  // The actual number of element in the vector which can be accessed.
+  std::size_t Size = 0;
+  // The position of the initial element of the page in the Data vector.
+  // Pages are allocated contiguously in the Data vector.
+  mutable std::vector<int> Lookup;
+  // Actual page data. All the page elements are added to this vector on the
+  // first access of any of the elements of the page. Elements default
+  // constructed and elements of the page are stored contiguously. The oder of
+  // the elements however depends on the order of access of the pages.
+  mutable std::vector<T> Data;
```

Comment:
Do you expect this storage to be >4G elements? If not, would it make sense to define it in terms of `SmallVector<T, 0>` or `SmallVector<T>`?

Reply:
At the moment the largest vector I found is 1.5M entries. Thinking about it, one could however in principle use this to allocate very sparsely placed pages (e.g. where the index is some 64bit spatial hash and elements tend to be clustered together). However, I do not have any (llvm) use case for that at the moment. I guess the real motivation was due to the fact I am not familiar with SmallVector<T> and I preferred to keep it simple.

Reply:
Using a single vector here seems unnecessarily inefficient, due to vector working to provide contiguous storage, which you don't need here.

It would be better to instead store pointers in your `PageToData` vector, and store a block of unallocated pages, allocating a new block (with exponential growth, like `vector`), when you run out. Alternatively, you could make this container take a `BumpPtrAllocator` for its pages, and pass in the Clang AST's allocator; that way you could share the same memory pool as the AST and minimize overallocation.

Reply:
I can try to use the allocator, indeed.